File: catkin_ws/src/edge_detection_ros/edge_detection/src/edge_detector.py
# import numpy as np 
# import cv2


# img = cv2.imread("/home/ros/neura/catkin_ws/src/edge_detection_ros/edge_detection/data/Image_2.png")

# img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

# img_blur = cv2.GaussianBlur(img_gray, (3,3), 0) 

# edges = cv2.Canny(image=img_blur, threshold1=100, threshold2=200, apertureSize=3)

# # kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (1, 1), anchor=(-1, -1))

# # dilated_edges = cv2.dilate(edges,kernel)

# # contours, _ = cv2.findContours(dilated_edges, cv2.RETR_LIST, cv2.CHAIN_APPROX_SIMPLE)

# cv2.imshow('Canny Edge Detection', edges)
# cv2.waitKey(0)
 
# cv2.destroyAllWindows()


import cv2
import numpy as np

import cv2

class EdgeDetector:
    def __init__(self, canny_threshold_1=50, canny_threshold_2=150, aperture_size=3, dilation_size=5):
        self.canny_threshold_1 = canny_threshold_1
        self.canny_threshold_2 = canny_threshold_2
        self.aperture_size = aperture_size
        self.dilation_size = dilation_size
    
    def detectEdges(self, image):
        # Convert the image to grayscale
        gray_image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

        # Apply a Gaussian blur to the image to remove noise
        gray_image = cv2.GaussianBlur(gray_image, (3, 3), 0)

        # Apply the Canny edge detection algorithm
        edges = cv2.Canny(gray_image, self.canny_threshold_1, self.canny_threshold_2, self.aperture_size)

        # Dilate the edges to fill in gaps
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (self.dilation_size, self.dilation_size))
        dilated_edges = cv2.dilate(edges, kernel)

        # Create a color image with green edges superimposed on the original image
        edges_image = cv2.cvtColor(dilated_edges, cv2.COLOR_GRAY2BGR)
        edges_image[np.where((edges_image == [255, 255, 255]).all(axis=2))] = [0, 255, 0]
        output_image = cv2.addWeighted(image, 0.7, edges_image, 0.3, 0)

        # # Get the dimensions of the image
        # height, width = output_image.shape[:2]

        # # Define the coordinates of the bounding box
        # x1, y1 = 0, 0
        # x2, y2 = width - 1, height - 1

        # # Draw the bounding box on the image
        # cv2.rectangle(output_image, (x1, y1), (x2, y2), (0, 255, 0), 2)

        # Return the color image with edges superimposed on the original image
        return output_image

# Create an instance of the EdgeDetector class with desired parameters
canny_threshold_1 = 300
canny_threshold_2 = 300
aperture_size = 3
dilation_size = 6
edge_detector = EdgeDetector(canny_threshold_1, canny_threshold_2, aperture_size, dilation_size)

# Load an image
image_path = "/home/ros/neura/catkin_ws/src/edge_detection_ros/edge_detection/data/Image_2.png"
image = cv2.imread(image_path)

# Detect edges and superimpose on the image
output_image = edge_detector.detectEdges(image)

# Display the output image
cv2.imshow("Edges Image", output_image)
cv2.waitKey(0)
cv2.destroyAllWindows()



# import cv2

# # Load an image
# image = cv2.imread('example_image.jpg')

# # Get the dimensions of the image
# height, width = image.shape[:2]

# # Define the coordinates of the bounding box
# x1, y1 = 0, 0
# x2, y2 = width - 1, height - 1

# # Draw the bounding box on the image
# cv2.rectangle(image, (x1, y1), (x2, y2), (0, 255, 0), 2)

# # Display the result
# cv2.imshow('Bounding box', image)
# cv2.waitKey(0)
# cv2.destroyAllWindows()

File: catkin_ws/src/edge_detection_ros/edge_detection/src/EdgeDetector.cpp
#include <edge_detection/EdgeDetector.hpp>
#include <iostream>

namespace edge_detection
{
    // Constructor
    EdgeDetector::EdgeDetector(double canny_threshold_1, double canny_threshold_2,
                               int aperture_size, int dilation_size)
        : canny_threshold_1_(canny_threshold_1), canny_threshold_2_(canny_threshold_2),
        aperture_size_(aperture_size), dilation_size_(dilation_size) 
    {}

    // Destructor
    EdgeDetector::~EdgeDetector() {}

    // Detect Edges
    cv::Mat EdgeDetector::detectEdges(cv::Mat& image)
    {
        // Convert the image to grayscale
        cv::Mat gray_image;
        cv::cvtColor(image, gray_image, cv::COLOR_BGR2GRAY);

        // Apply a Gaussian blur to the image to remove noise
        cv::GaussianBlur(gray_image, gray_image, cv::Size(3, 3), 0);

        // Apply the Canny edge detection algorithm
        cv::Mat edges;
        cv::Canny(gray_image, edges, canny_threshold_1_, canny_threshold_2_, aperture_size_);

        // Dilate the edges to fill in gaps
        cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, 
                                                   cv::Size(dilation_size_, dilation_size_));
        cv::Mat dilated_edges;
        cv::dilate(edges, dilated_edges, kernel);

        return dilated_edges;
    }

    // Color the Edges Green and Superimpose on Original Image
    cv::Mat EdgeDetector::ColorEdges(cv::Mat& image, cv::Mat& dilated_edges)
    {
        // Create a color image with green edges superimposed on the original image
        cv::Mat edges_image;
        cv::cvtColor(dilated_edges, edges_image, cv::COLOR_GRAY2BGR);
        for (int i = 0; i < edges_image.rows; i++) {
            for (int j = 0; j < edges_image.cols; j++) {
                if (edges_image.at<cv::Vec3b>(i, j) == cv::Vec3b(255, 255, 255)) {
                    edges_image.at<cv::Vec3b>(i, j) = cv::Vec3b(0, 255, 0);
                }
            }
        }

        cv::Mat output_image;
        cv::addWeighted(image, 0.7, edges_image, 0.3, 0, output_image);

        // Return the color image with edges superimposed on the original image
        return output_image;
    }
}

int main()
{
	edge_detection::EdgeDetector detector;
    
    // Create the executable for testing the code here

    // Load an input image from a file
    cv::Mat input_image = cv::imread("../data/Image_1.png");

    // Perform edge detection on the input image
    cv::Mat dilated_edges = detector.detectEdges(input_image);

    // Superimpose on input image
    cv::Mat output_image = detector.ColorEdges(input_image, dilated_edges);

    // Display the output image
    cv::imshow("Output Image", output_image);
    cv::waitKey(0);

	return 0;
}
